server/persistence/mysql/games.go

```diff
@@ -0,0 +1,571 @@
+		Player2ID VARCHAR(` + maxPlayerUUIDLenStr + `) NOT NULL,
+		Player3ID VARCHAR(` + maxPlayerUUIDLenStr + `),
+		Player4ID VARCHAR(` + maxPlayerUUIDLenStr + `),
+		PRIMARY KEY (GameID)
+	) ENGINE = INNODB;`
+
+	queryLatestGame = `SELECT 
+		gp.Player1ID, gp.Player2ID, gp.Player3ID, gp.Player4ID,
+		g.ScoreBlue, g.ScoreRed, g.ScoreGreen,
+		g.ScoreBlueLag, g.ScoreRedLag, g.ScoreGreenLag,
+		g.Phase, g.BlockingPlayers, g.CurrentDealer,
+		g.Hands, g.Crib, g.CutCard,
+		g.PeggedCards,
+		g.NumActions, g.Action
+	FROM Games g
+	INNER JOIN GamePlayers gp
+		ON g.GameID = gp.GameID
+		WHERE g.GameID = ? 
+	ORDER BY
+		NumActions DESC
+	LIMIT 1;`
+
+	queryGameAtNumActions = `SELECT 
+		gp.Player1ID, gp.Player2ID, gp.Player3ID, gp.Player4ID,
+		g.ScoreBlue, g.ScoreRed, g.ScoreGreen,
+		g.ScoreBlueLag, g.ScoreRedLag, g.ScoreGreenLag,
+		g.Phase, g.BlockingPlayers, g.CurrentDealer,
+		g.Hands, g.Crib, g.CutCard,
+		g.PeggedCards,
+		g.NumActions, g.Action
+	FROM Games g
+	INNER JOIN GamePlayers gp
+		ON g.GameID = gp.GameID
+	WHERE g.GameID = ? AND
+		g.NumActions = ?
+	;`
+
+	queryPlayerActionsBefore = `SELECT 
+		NumActions, Action
+	FROM Games
+	WHERE GameID = ? AND
+		NumActions <= ?
+	;`
+
+	addPlayersToGamePlayers = `INSERT INTO GamePlayers
+		(
+			GameID, 
+			Player1ID, Player2ID, Player3ID, Player4ID
+		)
+	VALUES
+		(
+			?,
+			?, ?, ?, ?
+		)
+	;`
+
+	insertGameAt = `INSERT INTO Games
+		(
+			GameID, NumActions, 
+			ScoreBlue, ScoreRed, ScoreGreen,
+			ScoreBlueLag, ScoreRedLag, ScoreGreenLag,
+			Phase, CutCard, Crib,
+			CurrentDealer,
+			BlockingPlayers, Hands, PeggedCards, Action
+		)
+	VALUES
+		(
+			?, ?,
+			?, ?, ?,
+			?, ?, ?,
+			?, ?, ?,
+			?,
+			?, ?, ?, ?
+		)
+	;`
+)
+
+var (
+	gamesCreateStmts = []string{
+		createGameTable,
+		createGamePlayersTable,
+	}
+)
+
+var _ persistence.GameService = (*gameService)(nil)
+
+type gameService struct {
+	db *txWrapper
+}
+
+func getGameService(
+	ctx context.Context,
+	db *txWrapper,
+) (persistence.GameService, error) {
+
+	for _, createStmt := range gamesCreateStmts {
+		_, err := db.ExecContext(ctx, createStmt)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &gameService{
+		db: db,
+	}, nil
+}
+
+func (g *gameService) Get(id model.GameID) (model.Game, error) {
+	r := g.db.QueryRow(queryLatestGame, id)
+	return g.populateGameFromRow(id, r)
+}
+
+func (g *gameService) GetAt(id model.GameID, numActions uint) (model.Game, error) {
+	r := g.db.QueryRow(queryGameAtNumActions, id, numActions)
+	return g.populateGameFromRow(id, r)
+}
+
+func (g *gameService) populateGameFromRow(
+	gID model.GameID,
+	r *sql.Row,
+) (model.Game, error) {
+
+	var p1ID, p2ID model.PlayerID
+	var p3ID, p4ID *model.PlayerID
+	var curDealerID model.PlayerID
+	var scoreBlue, scoreRed, scoreGreen,
+		lagScoreBlue, lagScoreRed, lagScoreGreen uint8
+	var phase model.Phase
+	var cribCardInts int32
+	var cutCardInt int8
+	var blockingPlayers, hands, peggedCards, action []byte
+	var numActions uint32
```

Comment:
Asking for a friend, would this be cleaner like this?
```go
var (
    ...
)
```

Reply:
🤷 either way works. the problem with a block is it puts the type so far away from the varnames:

```go
	var (
		p1ID, p2ID                               model.PlayerID
		p3ID, p4ID                               *model.PlayerID
		curDealerID                              model.PlayerID
		scoreBlue, scoreRed, scoreGreen          uint8
		lagScoreBlue, lagScoreRed, lagScoreGreen uint8
		phase                                    model.Phase
		cribCardInts                             int32
		cutCardInt                               int8
		blockingPlayers, hands                   []byte
		peggedCards, action                      []byte
		numActions                               uint32
	)
```